tslearn/early_classification.py

```diff
@@ -0,0 +1,312 @@
+
+    Parameters
+    ----------
+    n_clusters : int
+        Number of clusters to form.
+
+    base_classifier : Estimator
+        Estimator (instance) to be cloned and used for classifications.
+
+    minimum_time_stamp : int
+        Earliest time at which a classification can be performed on a time series
+
+    lamb : float
+        Value of the hyper parameter lambda used during the computation of the cost function to evaluate the probability
+        that a time series belongs to a cluster given the time series.
+
+    cost_time_parameter : float
+        Parameter of the cost function of time. This function is of the form : f(time) = time * cost_time_parameter
+
+    Attributes
+    --------------------
+    silhouette_ : float
+        The silhouette score from the clustering
+
+    classifiers_ : list
+        A list containing all the classifiers trained for the model
+
+    clusters_ : dictionary
+        Contains the times series by clusters
+
+    pyhatyck_ : dictionary
+        Contains the probabilities of being classified as class y_hat given class y and cluster ck
+
+    indice_ck_ : list
+        Contains for each clusters the indexes of the time series in the dataset
+
+    __n_classes_ : int private
+        The number of classes in the series
+
+    __len_X_ : int
+        The length of the time series
+
+    Examples
+    --------------------
+
+    References
+    --------------------
+    [1] A. Dachraoui, A. Bondu & A. Cornuéjols. Early classification of time series as a non myopic sequential decision
+    making problem. 2015 Conference paper
+
+
+    """
+
+    def __init__(
+        self,
+        n_clusters,
+        base_classifier,
+        minimum_time_stamp,
+        lamb,
+        cost_time_parameter,
+    ):
+        super(NonMyopicEarlyClassification, self).__init__()
+        self.base_classifier = base_classifier
+        self.n_clusters = n_clusters
+        self.minimum_time_stamp = minimum_time_stamp
+        self.lamb = lamb
+        self.cost_time_parameter = cost_time_parameter
+
+    def fit(self, X, y):
+        """
+        This function fits classifiers that are currently multilayer perceptrons to a training set of time series and
+        associated classes. A classifier is fit for each time stamp above a minimum time stamp that is an attribute of
+        the class of the model. Then some probabilities are computed using clusters already computed.
+        This function should be divided among 2 or 3 functions.
+
+        Parameters
+        ----------
+        X: Array-like
+            a dataset of time series
+        y: vector
+            the associated classes of the series from X_train
+        """
+        y_classes = np.unique(y)
+        self.labels_ = sorted(set(y_classes))
+        y_classes_indices = [self.labels_.index(yi) for yi in y_classes]
+
+        for idx, current_classe in enumerate(y_classes):
+            y[y == current_classe] = y_classes_indices[idx]
```

Comment:
You should never change `X` or `y` here. If you really need to do something like that, you should first copy `y` to a new `y_` and then operate on this new one. This is because the end-user never expects that calling fit for an estimator could change its data.